osu.Server.Spectator/Hubs/EntityStore.cs

```diff
@@ -0,0 +1,158 @@
+            {
+                if (!entityMapping.TryGetValue(id, out item))
+                    // was not tracking.
+                    return;
+            }
+
+            await item.ObtainLockAsync();
+
+            // handles removal and disposal of the semaphore.
+            item.Destroy();
+        }
+
+        /// <summary>
+        /// Get all tracked entities in an unsafe manner. Only read operations should be performed on retrieved entities.
+        /// </summary>
+        public KeyValuePair<long, T?>[] GetAllEntities()
+        {
+            lock (entityMapping)
+            {
+                return entityMapping
+                       .Where(kvp => kvp.Value.Item != null)
+                       .Select(entity => new KeyValuePair<long, T?>(entity.Key, entity.Value.Item))
+                       .ToArray();
+            }
+        }
+
+        /// <summary>
+        /// Clear all tracked entities.
+        /// </summary>
+        public void Clear()
+        {
+            lock (entityMapping)
+            {
+                entityMapping.Clear();
+            }
+        }
+
+        private void remove(long id)
+        {
+            lock (entityMapping)
+                entityMapping.Remove(id);
+        }
+
+        public class TrackedEntity
+        {
+            private readonly SemaphoreSlim semaphore = new SemaphoreSlim(1);
+
+            private T? item;
+
+            private readonly long id;
+            private readonly EntityStore<T> store;
+
+            private bool isDestroyed;
+
+            private bool isLocked => semaphore.CurrentCount == 0;
+
+            public TrackedEntity(long id, EntityStore<T> store)
+            {
+                this.id = id;
+                this.store = store;
+            }
+
+            public T? Item
+            {
+                get
+                {
+                    checkValidForUse();
+                    return item;
+                }
+                set
+                {
+                    checkValidForUse();
+                    item = value;
+                }
+            }
+
+            /// <summary>
+            /// Mark this item as no longer used. Will remove any tracking overhead.
+            /// </summary>
+            public void Destroy()
+            {
+                // we should already have a lock when calling destroy.
+                checkValidForUse();
+
+                store.remove(id);
+                semaphore.Dispose();
+                isDestroyed = true;
+            }
+
+            public async Task ObtainLockAsync()
+            {
+                checkValidForUse(false);
+
+                if (!await semaphore.WaitAsync(lock_timeout))
+                    throw new TimeoutException($"Lock for {typeof(T)} id {id} could not be obtained within timeout period");
+            }
+
+            public void ReleaseLock()
+            {
+                if (!isDestroyed)
+                    semaphore.Release();
```

Comment:
Can you explain what you mean here? What would a hard fail entail?

I'd hope this method is only ever called by the `EntityStore` but enforcing that isn't so simple without making interfaces or otherwise.

Reply:
I just meant that this is the only public method of `TrackedEntity` (aside from `GetItemUnsafe()`) that does not call `checkValidForUse()` as a precondition before proceeding. Again not a huge deal.